smart-contracts/assembly/contracts/NFT/NFT-internals.ts

```diff
@@ -1,282 +1,239 @@
-  assert(_onlyMinted(tokenId), `Token ${tokenId.toString()} is not minted`);
+/**
+ * Returns the allowance set on the owner's account for the spender.
+ *
+ * @param operator - address of the operator
+ * @param tokenId - tokenId of the token
+ * @returns
+ */
+export function _isApproved(operator: string, tokenId: u256): bool {
+  const allowKey = allowanceKey(tokenId);
+  if (Storage.has(allowKey)) {
+    return bytesToString(Storage.get(allowKey)) == operator;
+  }
+  return false;
 }
 
-export function assertIsOwner(address: string, tokenId: u64): void {
-  assert(
-    _isTokenOwner(address, tokenId),
-    `${address} is not the owner of ${tokenId.toString()}`,
-  );
+/**
+ * Enable or disable approval for a third party ("operator") to manage all of `Context.caller`'s assets
+ *
+ * @param operator - Address to add to the set of authorized operators
+ * @param approved - True if the operator is approved, false to revoke approval
+ */
+export function _setApprovalForAll(operator: string, approved: bool): void {
+  const key = operatorAllowanceKey(Context.caller().toString(), operator);
+  if (approved) {
+    Storage.set(key, boolToByte(true));
+  } else {
+    Storage.del(key);
+  }
 }
 
-function assertIsApproved(owner: string, caller: string, tokenId: u64): void {
-  assert(
-    _isApproved(caller, tokenId) ||
-      _isApprovedForAll(owner, caller) ||
-      owner === caller,
-    `${caller} is not allowed to transfer this token`,
-  );
+/**
+ * Query if an address is an authorized operator for another address
+ * @param owner - The address that owns the NFTs
+ * @param operator - The address that acts on behalf of the owner
+ * @returns
+ */
+export function _isApprovedForAll(owner: string, operator: string): bool {
+  const key = operatorAllowanceKey(owner, operator);
+  if (Storage.has(key)) {
+    return byteToBool(Storage.get(key));
+  }
+  return false;
 }
 
-export function assertNotSelfTransfer(owner: string, recipient: string): void {
-  assert(
-    owner != recipient,
-    'The owner and the recipient must be different addresses',
+/**
+ * Returns whether operator is allowed to manage tokenId.
+ *
+ * @param operator - address of the operator
+ * @param tokenId - The NFT to be managed
+ * @returns
+ */
+function _isAuthorized(operator: string, tokenId: u256): bool {
+  return (
+    _isApproved(operator, tokenId) ||
+    _isApprovedForAll(_ownerOf(tokenId), operator) ||
+    _ownerOf(tokenId) == operator
   );
 }
 
-export function assertAddressIsValid(address: string): void {
-  assert(validateAddress(address), 'Address is not valid');
+export function _update(to: string, tokenId: u256, auth: string): void {
+  const from = _ownerOf(tokenId);
+  if (auth != '') {
+    assert(_isAuthorized(auth, tokenId), 'Unauthorized');
+  }
+  if (from != '') {
+    // clear the approval
+    _approve('', tokenId);
+    // update the balance of the from
+    const fromBalance = bytesToU256(Storage.get(balanceKey(from)));
```

Comment:
You're not checking that the balance is not 0

Reply:
Balance cannot be zero because from is `_ownerOf(tokenId)`